controllers/metal3.io/bmceventsubscription_controller.go

```diff
@@ -0,0 +1,260 @@
+		Namespace: request.Namespace,
+	}
+	err = r.Get(ctx, namespacedHostName, host)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			reqLogger.Error(err, "baremetalhost not found")
+			return ctrl.Result{RequeueAfter: subscriptionRetryDelay}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, errors.Wrap(err, "could not load host data")
+	}
+
+	// Add a finalizer to newly created objects.
+	if subscription.DeletionTimestamp.IsZero() && !subscriptionHasFinalizer(subscription) {
+		reqLogger.Info(
+			"adding finalizer",
+			"existingFinalizers", subscription.Finalizers,
+			"newValue", metal3v1alpha1.BMCEventSubscriptionFinalizer,
+		)
+		subscription.Finalizers = append(subscription.Finalizers,
+			metal3v1alpha1.BMCEventSubscriptionFinalizer)
+		err := r.Update(ctx, subscription)
+		if err != nil {
+			return ctrl.Result{}, errors.Wrap(err, "failed to add finalizer")
+		}
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	// Retrieve the BMC details from the host spec and validate host
+	// BMC details and build the credentials for talking to the
+	// management controller.
+	var bmcCreds *bmc.Credentials
+
+	bmcCreds, _, err = r.buildAndValidateBMCCredentials(request, host)
+	if err != nil || bmcCreds == nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to get bmc secret")
+	}
+
+	prov, err := r.ProvisionerFactory.NewProvisioner(provisioner.BuildHostData(*host, *bmcCreds), nil)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to create provisioner")
+	}
+
+	ready, err := prov.IsReady()
+	if err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to check services availability")
+	}
+	if !ready {
+		reqLogger.Info("provisioner is not ready", "RequeueAfter:", provisionerNotReadyRetryDelay)
+		return ctrl.Result{Requeue: true, RequeueAfter: provisionerNotReadyRetryDelay}, nil
+	}
+
+	httpHeaders, err := r.getHTTPHeaders(*subscription)
+
+	if err != nil {
+		reqLogger.Error(err, "failed to get http headers")
+		return ctrl.Result{RequeueAfter: subscriptionRetryDelay}, nil
+	}
+
+	if subscription.DeletionTimestamp.IsZero() {
+		// Not being deleted
+		if subscription.Status.SubscriptionID == "" {
+			if _, err := prov.AddBMCEventSubscriptionForNode(subscription, httpHeaders); err != nil {
+				return ctrl.Result{}, errors.Wrap(err, "failed to add a subscription")
+			}
+
+			// Update the subscription with the new ID
+			err := r.Status().Update(ctx, subscription)
+			if err != nil {
+				return ctrl.Result{}, errors.Wrap(err, "failed to update subscription status")
+			}
+
+		} else {
+			// Update an existing subscription by recreating it in the BMC
+
+			originalSubscription := subscription.DeepCopy()
+
+			if _, err := prov.AddBMCEventSubscriptionForNode(subscription, httpHeaders); err != nil {
+				return ctrl.Result{}, errors.Wrap(err, "failed to update subscription")
+			}
+
+			if _, err := prov.RemoveBMCEventSubscriptionForNode(*originalSubscription); err != nil {
```

Comment:
I don't know how many simultaneous subscriptions a BMC allows, I'd rather delete the old subscription *before* creating a new one